mllib/src/main/scala/org/apache/spark/ml/evaluation/RankingEvaluator.scala

```diff
@@ -0,0 +1,287 @@
+    require(predictionType == labelType,
+      s"Prediction column $predictionColName and Label column $labelColName " +
+        s"must be of the same type, but Prediction column $predictionColName is $predictionType " +
+        s"and Label column $labelColName is $labelType")
+
+    val metric = $(metricName) match {
+      case "map" => meanAveragePrecision(dataset)
+      case "ndcg" => normalizedDiscountedCumulativeGain(dataset)
+      case "mapk" => meanAveragePrecisionAtK(dataset)
+      case "mrr" => meanReciprocalRank(dataset)
+    }
+    metric
+  }
+
+  /**
+   * Returns the mean average precision (MAP) of all the queries.
+   * If a query has an empty ground truth set, the average precision will be zero and a log
+   * warning is generated.
+   */
+  private def meanAveragePrecision(dataset: Dataset[_]): Double = {
+    val sc = SparkContext.getOrCreate()
+    val sqlContext = SQLContext.getOrCreate(sc)
+    import sqlContext.implicits._
+
+    dataset.map{ case (prediction: Array[T], label: Array[T]) =>
+      val labSet = label.toSet
+
+      if (labSet.nonEmpty) {
+        var i = 0
+        var cnt = 0
+        var precSum = 0.0
+        val n = prediction.length
+        while (i < n) {
+          if (labSet.contains(prediction(i))) {
+            cnt += 1
+            precSum += cnt.toDouble / (i + 1)
+          }
+          i += 1
+        }
+        precSum / labSet.size
+      } else {
+        logWarning("Empty ground truth set, check input data")
+        0.0
+      }
+    }.reduce{ (a, b) => a + b } / dataset.count
+  }
+
+  /**
+   * Compute the average NDCG value of all the queries, truncated at ranking position k.
+   * The discounted cumulative gain at position k is computed as:
+   *    sum,,i=1,,^k^ (2^{relevance of ''i''th item}^ - 1) / log(i + 1),
+   * and the NDCG is obtained by dividing the DCG value on the ground truth set. In the current
+   * implementation, the relevance value is binary.
+
+   * If a query has an empty ground truth set, zero will be used as ndcg together with
+   * a log warning.
+   *
+   * See the following paper for detail:
+   *
+   * IR evaluation methods for retrieving highly relevant documents. K. Jarvelin and J. Kekalainen
+   */
+  private def normalizedDiscountedCumulativeGain(dataset: Dataset[_]): Double = {
+    val sc = SparkContext.getOrCreate()
+    val sqlContext = SQLContext.getOrCreate(sc)
+    import sqlContext.implicits._
+
+    dataset.map{ case (prediction: Array[T], label: Array[T]) =>
+      val labSet = label.toSet
+
+      if (labSet.nonEmpty) {
+        val labSetSize = labSet.size
+        val n = math.min(math.max(prediction.length, labSetSize), $(k))
+        var maxDcg = 0.0
+        var dcg = 0.0
+        var i = 0
+        while (i < n) {
+          val gain = 1.0 / math.log(i + 2)
+          if (labSet.contains(prediction(i))) {
+            dcg += gain
+          }
+          if (i < labSetSize) {
+            maxDcg += gain
```

Comment:
Relevance here is 0/1. It doesn't matter if you calculate the ideal DCG to the number of all relevant items, or all the way to n, since everything after the number of relevant items has relevance 0 and does not add to the sum.